src/ast.rs

```diff
@@ -6021,7 +6015,8 @@ where
             } else {
                 start
             };
-            ctx.write_all(&ctx.input[start..end])
+            let s = ::std::str::from_utf8(&ctx.input[start..end]).map_err(|_| fmt::Error)?;
```

Comment:
Is this `.map_err(|_| fmt::Error)` okay? I'm doing this in multiple places in this PR. I could also create a custom error type or use the error type from the the error module of this crate (extending it with `fmt::Error` support previously).